tests/iio_writedev.c

```diff
@@ -0,0 +1,437 @@
+		return EXIT_FAILURE;
+	}
+
+	if (timeout >= 0)
+		iio_context_set_timeout(ctx, timeout);
+
+	dev = iio_context_find_device(ctx, argv[optind]);
+	if (!dev) {
+		fprintf(stderr, "Device %s not found\n", argv[optind]);
+		iio_context_destroy(ctx);
+		return EXIT_FAILURE;
+	}
+
+	if (trigger_name) {
+		struct iio_device *trigger = iio_context_find_device(
+				ctx, trigger_name);
+		if (!trigger) {
+			fprintf(stderr, "Trigger %s not found\n", trigger_name);
+			iio_context_destroy(ctx);
+			return EXIT_FAILURE;
+		}
+
+		if (!iio_device_is_trigger(trigger)) {
+			fprintf(stderr, "Specified device is not a trigger\n");
+			iio_context_destroy(ctx);
+			return EXIT_FAILURE;
+		}
+
+		/*
+		 * Fixed rate for now. Try new ABI first,
+		 * fail gracefully to remain compatible.
+		 */
+		if (iio_device_attr_write_longlong(trigger,
+				"sampling_frequency", DEFAULT_FREQ_HZ) < 0)
+			iio_device_attr_write_longlong(trigger,
+				"frequency", DEFAULT_FREQ_HZ);
+
+		iio_device_set_trigger(dev, trigger);
+	}
+
+	nb_channels = iio_device_get_channels_count(dev);
+
+	if (argc == optind + 1) {
+		/* Enable all channels */
+		for (i = 0; i < nb_channels; i++)
+			iio_channel_enable(iio_device_get_channel(dev, i));
+	} else {
+		for (i = 0; i < nb_channels; i++) {
+			unsigned int j;
+			struct iio_channel *ch = iio_device_get_channel(dev, i);
+			for (j = optind + 1; j < (unsigned int) argc; j++) {
+				const char *n = iio_channel_get_name(ch);
+				if (!strcmp(argv[j], iio_channel_get_id(ch)) ||
+						(n && !strcmp(n, argv[j])))
+					iio_channel_enable(ch);
+			}
+		}
+	}
+
+	sample_size = iio_device_get_sample_size(dev);
+
+	buffer = iio_device_create_buffer(dev, buffer_size, cyclic_buffer);
+	if (!buffer) {
+		char buf[256];
+		iio_strerror(errno, buf, sizeof(buf));
+		fprintf(stderr, "Unable to allocate buffer: %s\n", buf);
+		iio_context_destroy(ctx);
+		return EXIT_FAILURE;
+	}
+
+	while (app_running) {
+		/* If there are only the samples we requested, we don't need to
+		 * demux */
+		if (iio_buffer_step(buffer) == sample_size) {
+			void *start = iio_buffer_start(buffer);
+			size_t write_len, len = (intptr_t) iio_buffer_end(buffer)
+				- (intptr_t) start;
+
+			if (num_samples && len > num_samples * sample_size)
+				len = num_samples * sample_size;
+
+			for (write_len = len; len; ) {
+				size_t nb = fread(start, 1, len, stdin);
+				if (!nb)
+					goto err_destroy_buffer;
+
+				len -= nb;
+				start = (void *)((intptr_t) start + nb);
+			}
+
+			if (num_samples) {
+				num_samples -= write_len / sample_size;
+				if (!num_samples)
+					quit_all(EXIT_SUCCESS);
+			}
+		} else {
+			iio_buffer_foreach_sample(buffer, read_sample, NULL);
+		}
+
+		int ret = iio_buffer_push(buffer);
+		if (ret < 0) {
+			if (app_running) {
+				char buf[256];
+				iio_strerror(-ret, buf, sizeof(buf));
+				fprintf(stderr, "Unable to push buffer: %s\n", buf);
+			}
+			break;
+		}
+
+		while(cyclic_buffer && app_running)
```

Comment:
this looks like it would hang indefinitely here, without any operation;
that is, until `app_running` is false via sig-handler

is this intended ?

Reply:
i did not dig here yet;
but does the IIO lib have some internal threading that would do operation here ?

Reply:
Yes - a cyclic buffer is only pushed once. This ensures that the buffer remains running until a signal is catched. Using sleep here is beneficial since it works cross platform equally well.